stumpy/gpu_stump.py

```diff
@@ -354,27 +416,52 @@ def _gpu_stump(
                 device_Σ_T,
                 device_μ_Q,
                 device_σ_Q,
-                k,
+                profile_len,
                 ignore_trivial,
                 excl_zone,
                 device_profile,
+                device_profile_L,
+                device_profile_R,
                 device_indices,
+                device_indices_L,
+                device_indices_R,
                 True,
+                bfs,
+                nlevel,
+                k,
             )
 
         profile = device_profile.copy_to_host()
+        profile_L = device_profile_L.copy_to_host()
+        profile_R = device_profile_R.copy_to_host()
         indices = device_indices.copy_to_host()
+        indices_L = device_indices_L.copy_to_host()
+        indices_R = device_indices_R.copy_to_host()
+
         profile = np.sqrt(profile)
+        profile_L = np.sqrt(profile_L)
+        profile_R = np.sqrt(profile_R)
 
         profile_fname = core.array_to_temp_file(profile)
+        profile_L_fname = core.array_to_temp_file(profile_L)
+        profile_R_fname = core.array_to_temp_file(profile_R)
         indices_fname = core.array_to_temp_file(indices)
+        indices_L_fname = core.array_to_temp_file(indices_L)
+        indices_R_fname = core.array_to_temp_file(indices_R)
 
-    return profile_fname, indices_fname
+    return (
+        profile_fname,
+        profile_L_fname,
+        profile_R_fname,
+        indices_fname,
+        indices_L_fname,
+        indices_R_fname,
+    )
 
 
 @core.non_normalized(gpu_aamp)
 def gpu_stump(
```

Comment:
Did you happen to test if `gpu_stump` is slower for different `k`?

Reply:
Unfortunately, I did not test it. I will test old version and new version of `gpu_stump` for `k=1`, and also test new version for `k>1`.

Reply:
I tested the `top-k gpu_stump` and I think the computing time has increased significantly! I think I need to figure out what part(s) cause such increase in computing time.
```
seed = 0
np.random.seed(seed)
T = np.random.rand(50000)
m = 50
```

|  k=1 | stump | gpu_stump |
| ------------- | ------| ------- |
| original version  | 3.1 | 11  |
| top-k version | 3.7 | 45  |

Also:
|  k=10 | stump | gpu_stump |
| ------------- | ------| ------- |
| top-k version | 13.7 | 47 |

Reply:
@seanlaw 
**Update**
I forgot to copy array `bfs` to device and thus there was `copy overhead to device`. Now, after fixing the issue and passing the device parameter to the kernel, the time is around  `16` seconds.

Reply:
Excellent. These numbers are pretty good. So, is it around 16 seconds for BOTH `k=1` and `k=10` for `T=50_000`?

Reply:
> 16 seconds for BOTH `k=1` and `k=10` for `T=50_000`?

All computing time for `k=1` to `k=10` was about 16~18 sec, and it was not monotonic. In some cases, higher value of `k` has lower computing time. That is why I decided to provide an approximate value. I can check it again just to make sure of it.

> Hopefully, you can understand why I asked you to test the performance both times. These simple checks can reveal a lot!

Yes... I can see it! Monitoring the impact of changes on the performance can help us detect the parts that result in strange behavior. Although I think the structure of `stumped` is somehow similar to `stump` and there should be no problem, I will check the performance of `stumped` as well and will let you know if anything goes wrong.

Reply:
>  In other words, there is little to no overhead for 1 <= k <= 10?

![image](https://user-images.githubusercontent.com/38519522/171066082-58d4c85b-d869-43da-a192-461c834bfe0b.png)

So, there are some fluctuations...but overall, the performance is good!

Reply:
Hmm, that's really interesting. I wonder if the high number of threads is really helping a lot here (as compared with `stump` where the performance kept getting worse as `k` increased). Also, I wonder how much of this time variance is due to copying things over to the GPU. If possible, can you do the same thing but time the call to `_gpu_stump` instead (i.e., the line in `gpu_stump` that starts with `profile[idx], indices[idx] = _gpu_stump(`? I would expect this call to be relatively linear?

In case it matters, with CPU timing, it is rather important to not do anything else while it is executing (like performing other tasks) as it can have large effect. Just something to keep in mind.

Reply:
> In case it matters, with CPU timing, it is rather important to not do anything else while it is executing (like performing other tasks) as it can have large effect. Just something to keep in mind.

I did not know that! Thanks for the info. 

> If possible, can you do the same thing but time the call to _gpu_stump instead (i.e., the line in gpu_stump that starts with profile[idx], indices[idx] = _gpu_stump(? I would expect this call to be relatively linear?

I will update you after getting the result

Reply:
![image](https://user-images.githubusercontent.com/38519522/171074991-ae718200-08a5-4c6a-9dd1-d848a0ca30a6.png)

I added `tic=time.time()` [in this  line](https://github.com/TDAmeritrade/stumpy/blob/998e5d2017ed3d01d1a3993ce01e96de3dfd5254/stumpy/gpu_stump.py#L557), and did `toc=time.time()` at end of if-block. I also ran the code for 10 values of k in each time otherwise my pc becomes very HOT and I wasn't sure if it affects the performance or not. So, I tried it every 10 values to keep it cool.

Reply:
Thanks for the tests. Hmm... I wonder why there are two peaks at around `k=16` and `k=35` in both plots

Reply:
Also, I tested again, and considered timing of kernel function as well this time.

![image](https://user-images.githubusercontent.com/38519522/171081125-55987a6a-9cc2-4b6f-aeb7-cb54cc2bd0c6.png)

Here, the peaks have changed! It is now at `k=2` and `k=37`.

Reply:
Does this imply that there is more calls to '_gpu_searchsorted_right' where the peaks are?

Reply:
That can be the reason. But, I did not change the seed when generating random values for `T`. Each thread is responsible for  a particular set of subsequences, and I think we are not changing the subsequences assigned to each thread. Right? So, shouldn't the peaks stay the same?

Reply:
Yeah, I would also expect the peaks to stay the same or, at least, roughly the same.

Reply:
In below, the computing time for each value of `k` is the average of computing time on 11 different `T` (generated by 11 different seeds). I used the same set of seeds for all values of `k` to make sure the comparison is fair.

![image](https://user-images.githubusercontent.com/38519522/171093223-7c1be321-405c-4018-8506-0699e7d39864.png)

By the way, I ran the functions for `k=1 to 10` and then, after making sure my PC is not hot again, I ran for `k=11 to 20`. Now, if you look at the first half, and the second half, we can see both graphs are increasing. 

Two possible things:
(1) PC gets hot and it might affect the performance of transistors!
(2) Maybe GPU memory/threads need to be reset/free up after each usage (?!?)

Reply:
Hmm, honestly, I don't know what to conclude here. Maybe heat has something to do with differences in performance. I can't really tell. Anyhow, I think we'll have to be okay with it for now.